src/payment/strategies/square/square-payment-strategy.ts

```diff
@@ -46,34 +74,43 @@ export default class SquarePaymentStrategy extends PaymentStrategy {
         const { payment, ...order } = payload;
 
         if (!payment || !payment.methodId) {
-            throw new InvalidArgumentError('Unable to submit payment because "payload.payment.methodId" argument is not provided.');
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
         }
-
         const paymentName = payment.methodId;
-
-        return new Promise<NonceInstrument>((resolve, reject) => {
-            if (!this._paymentForm) {
-                throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
-            }
-
-            if (this._deferredRequestNonce) {
-                this._deferredRequestNonce.reject(new TimeoutError());
-            }
-
-            this._deferredRequestNonce = { resolve, reject };
-            this._paymentForm.requestCardNonce();
-        })
-        .then(paymentData => {
-            const paymentPayload = {
+        if (payment.paymentData && isNonceLike(payment.paymentData)) {
+            const paymentPayload: Payment = {
                 methodId: paymentName,
-                paymentData,
+                paymentData: payment.paymentData as NonceInstrument,
             };
-
+            const methodId = payload.payment;
             return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
                 .then(() =>
                     this._store.dispatch(this._paymentActionCreator.submitPayment(paymentPayload))
                 );
-        });
+        } else {
+            return new Promise<NonceInstrument>((resolve, reject) => {
+                if (!this._paymentForm) {
+                    throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+                }
+
+                if (this._deferredRequestNonce) {
+                    this._deferredRequestNonce.reject(new TimeoutError());
+                }
+
+                this._deferredRequestNonce = { resolve, reject };
+                this._paymentForm.requestCardNonce();
+            })
+            .then(paymentData => {
+                const paymentPayload = {
+                    methodId: paymentName,
+                    paymentData,
+                };
+                return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
```

Comment:
Since this is repeated twice:

```js
return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
    .then(() =>
        this._store.dispatch(this._paymentActionCreator.submitPayment(paymentPayload))
    );
```

It might look better if we create a method to retrieve the `paymentData`:


```ts
execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
    const { payment, ...order } = payload;

    if (!payment || !payment.methodId) {
        throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
    }

    const { methodId } = payment;

    return this._getPaymentData(payment)
        .then(paymentData => {
            const paymentPayload = { methodId, paymentData };

            return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
                .then(() =>
                    this._store.dispatch(this._paymentActionCreator.submitPayment(paymentPayload))
                );
        });
}
```

Reply:
👍